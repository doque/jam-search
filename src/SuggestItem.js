import React, { Component } from "react";

class SuggestItem extends Component {
  render() {
    const { header, list } = this.props;

    return (
      <div>
        <h2>{header}</h2>
        <ul>{list.map(item => <li>{item.term}</li>)}</ul>
      </div>
    );
  }
}

export default SuggestItem;
