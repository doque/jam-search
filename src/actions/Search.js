import { fetchSuggestions } from "../util/Api";

export const ENTER_SEARCHTERM = "ENTER_SEARCHTERM";
export const REQUEST_SUGGESTIONS = "REQUEST_SUGGESTIONS";
export const RECEIVE_SUGGESTIONS = "RECEIVE_SUGGESTIONS";
export const OFFLINE = "OFFLINE";

export const enteredSearchTerm = (dispatch, searchTerm) => {
  dispatch(requestedSuggestions(dispatch, searchTerm));

  return {
    type: ENTER_SEARCHTERM,
    searchTerm
  };
};

export const requestedSuggestions = (dispatch, searchTerm) => {
  fetchSuggestions(searchTerm)
    .then(json => {
      dispatch(receivedSuggestions(json.suggests));
      dispatch(offline(false));
    })
    .catch(e => {
      dispatch(offline(true));
    });

  return {
    type: REQUEST_SUGGESTIONS
  };
};

export const offline = isOffline => {
  return {
    type: OFFLINE,
    isOffline
  };
};

export const receivedSuggestions = suggestions => {
  return {
    type: RECEIVE_SUGGESTIONS,
    suggestions
  };
};
