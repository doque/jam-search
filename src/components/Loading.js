import React from "react";

const Loading = () =>
  <img src={`${process.env.PUBLIC_URL}/loading.svg`} alt="Loading..." />;

export default Loading;
