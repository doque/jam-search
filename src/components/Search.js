import React, { Component } from "react";

import Suggestions from "./Suggestions";
import SearchForm from "./SearchForm";

class SearchContainer extends Component {
  state = {
    data: {
      suggests: []
    },
    searchTerm: ""
  };

  changeHandler(event) {
    const value = event.target.value.trim();
    const type = event.target.name;

    if (value.length >= 3) {
      this.fetchData(value, type).then(data => this.setState({ data }));
    } else {
      this.setState({ data: { suggests: [] } });
    }
    this.setState({ searchTerm: value });
  }

  clickHandler(searchTerm) {
    this.setState({ searchTerm });
  }

  fetchData(value, type = "what") {
    const url = `https://jameda.localtunnel.me/suche.jameda-elements.de/${type}-new?query=${value}`;

    return fetch(url)
      .then(response => response.json())
      .catch(e => console.error("Probably offline?", e));
  }

  render() {
    return (
      <div>
        <SearchForm
          searchTerm={this.state.searchTerm}
          changeHandler={this.changeHandler.bind(this)}
        />
        <Suggestions
          data={this.state.data}
          clickHandler={this.clickHandler.bind(this)}
        />
      </div>
    );
  }
}

export default SearchContainer;
