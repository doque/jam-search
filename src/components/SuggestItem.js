import React from "react";
import PropTypes from "prop-types";

const SuggestItem = ({ header, list, clickHandler }) =>
  <div>
    <h2>
      {header}
    </h2>
    <ul>
      {list.map((item, i) =>
        <li onClick={() => clickHandler(item.term)} key={i}>
          {item.term}
        </li>
      )}
    </ul>
  </div>;

SuggestItem.propTypes = {
  header: PropTypes.string,
  list: PropTypes.array,
  clickHandler: PropTypes.func
};

export default SuggestItem;
