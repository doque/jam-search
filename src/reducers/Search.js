import {
  ENTER_SEARCHTERM,
  REQUEST_SUGGESTIONS,
  RECEIVE_SUGGESTIONS,
  OFFLINE
} from "../actions/Search";

const initialState = {
  suggestions: [],
  isFetching: false,
  searchTerm: "",
  isOffline: false
};

const Search = (state = initialState, action) => {
  switch (action.type) {
    case ENTER_SEARCHTERM:
      return {
        ...state,
        searchTerm: action.searchTerm
      };
    case REQUEST_SUGGESTIONS:
      return {
        ...state,
        isFetching: true
      };
    case RECEIVE_SUGGESTIONS:
      return {
        ...state,
        isFetching: false,
        suggestions: action.suggestions
      };
    case OFFLINE:
      return {
        ...state,
        isOffline: action.isOffline
      };
    default: {
      return state;
    }
  }
};

export default Search;
