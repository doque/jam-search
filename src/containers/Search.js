import React from "react";
import { connect } from "react-redux";

import { enteredSearchTerm } from "../actions/Search";

import Suggestions from "../components/Suggestions";
import SearchForm from "../components/SearchForm";
import Loading from "../components/Loading";
import Offline from "../components/Offline";

const Search = ({
  searchTerm,
  isFetching,
  suggestions,
  changeHandler,
  isOffline
}) =>
  <div>
    <SearchForm searchTerm={searchTerm} changeHandler={changeHandler} />
    {!isOffline && isFetching && <Loading />}
    {isOffline && <Offline />}
    <Suggestions suggestions={suggestions} />
  </div>;

const mapStateToProps = state => state;

const mapDispatchToProps = dispatch => ({
  changeHandler: searchTerm => dispatch(enteredSearchTerm(dispatch, searchTerm))
});

export default connect(mapStateToProps, mapDispatchToProps)(Search);
export { Search };
