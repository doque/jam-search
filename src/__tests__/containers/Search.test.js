import React from "react";
import ReactDOM from "react-dom";
import { render } from "enzyme";
import renderer from "react-test-renderer";

import { Search } from "../../containers/Search";
import { initialState } from "../../reducers/Search";
import Loading from "../../components/Loading";
import SuggestItem from "../../components/SuggestItem";

const defaultProps = initialState;

const renderWithProps = (props = defaultProps) => {
  return renderer.create(<Search {...props} />).toJSON();
};

it("renders without crashing", () => {
  const wrapper = renderWithProps();
  expect(wrapper).toMatchSnapshot();
});

it("renders the loading spinner when isFetching is true", () => {
  const wrapper = renderWithProps({
    ...defaultProps,
    isFetching: true
  });
  expect(wrapper).toMatchSnapshot();
});

it("does not render the loading spinner when isFetching is false", () => {
  const wrapper = renderWithProps({
    ...defaultProps,
    isFetching: false
  });

  expect(wrapper).toMatchSnapshot();
});

it("renders Offline when isOffline is true", () => {
  const wrapper = renderWithProps({
    ...defaultProps,
    isOffline: true
  });

  expect(wrapper).toMatchSnapshot();
});

it("does not render Offline when isOffline is false", () => {
  const wrapper = renderWithProps({
    ...defaultProps,
    isOffline: false
  });
});

it("renders the correct number of SuggestItems", () => {
  const data = require("../../__mocks__/Search");
  const wrapper = renderWithProps({ suggestions: data.suggests });
  expect(wrapper).toMatchSnapshot();
});
