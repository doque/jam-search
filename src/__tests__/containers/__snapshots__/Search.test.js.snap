// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render the loading spinner when isFetching is false 1`] = `
<div>
  <div>
    <label
      htmlFor="what"
    >
      Was:
      <input
        className="input-field"
        name="what"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <div />
</div>
`;

exports[`renders Offline when isOffline is true 1`] = `
<div>
  <div>
    <label
      htmlFor="what"
    >
      Was:
      <input
        className="input-field"
        name="what"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <div />
  <div>
    Offline
  </div>
</div>
`;

exports[`renders the correct number of SuggestItems 1`] = `
<div>
  <div>
    <label
      htmlFor="what"
    >
      Was:
      <input
        className="input-field"
        name="what"
        onChange={[Function]}
        type="text"
        value={undefined}
      />
    </label>
  </div>
  <div>
    <div>
      <h2>
        Fachbereiche
      </h2>
      <ul>
        <li
          onClick={[Function]}
        >
          Zahnarzt
        </li>
        <li
          onClick={[Function]}
        >
          Zahnärzte
        </li>
        <li
          onClick={[Function]}
        >
          Zahnärztin
        </li>
        <li
          onClick={[Function]}
        >
          Zahnmediziner
        </li>
        <li
          onClick={[Function]}
        >
          Zahnmedizinerin
        </li>
        <li
          onClick={[Function]}
        >
          Zahn-Klinik
        </li>
      </ul>
    </div>
    <div>
      <h2>
        Namen
      </h2>
      <ul>
        <li
          onClick={[Function]}
        >
          Zahn, Erika
        </li>
        <li
          onClick={[Function]}
        >
          Zahn, Ralf
        </li>
        <li
          onClick={[Function]}
        >
          Zahn, Pia
        </li>
        <li
          onClick={[Function]}
        >
          Zahn, Tino
        </li>
        <li
          onClick={[Function]}
        >
          Zahn, Tanja
        </li>
        <li
          onClick={[Function]}
        >
          Zahn, Axel
        </li>
      </ul>
    </div>
    <div>
      <h2>
        Symptome, Keywords, passende Fachbereiche
      </h2>
      <ul>
        <li
          onClick={[Function]}
        >
          Zahn
        </li>
        <li
          onClick={[Function]}
        >
          Zahnen
        </li>
        <li
          onClick={[Function]}
        >
          Zahnen
        </li>
        <li
          onClick={[Function]}
        >
          Zahnhals
        </li>
        <li
          onClick={[Function]}
        >
          Zahnloch
        </li>
        <li
          onClick={[Function]}
        >
          Zahnnerv
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`renders the loading spinner when isFetching is true 1`] = `
<div>
  <div>
    <label
      htmlFor="what"
    >
      Was:
      <input
        className="input-field"
        name="what"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <img
    alt="Loading..."
    src="loading.svg"
  />
  <div />
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  <div>
    <label
      htmlFor="what"
    >
      Was:
      <input
        className="input-field"
        name="what"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <div />
</div>
`;
